gateway/src/proxy.rs

```diff
@@ -109,9 +113,37 @@ async fn proxy(
             ip,
             std::time::Duration::from_millis(1000),
         );
+
+        if was_stopped {
+            // wait until service has started and opens its port, give up after 10s
+            span.in_scope(|| {
+                debug!("project waking up, checking service port");
+            });
+            let addr = SocketAddr::new(ip, DEPLOYER_SERVICE_HTTP_PORT);
+            let span_clone = span.clone();
+            let _ = tokio::time::timeout(Duration::from_secs(10), async move {
+                let mut ms = 5;
+                loop {
+                    if let Ok(socket) = TcpSocket::new_v4() {
+                        if socket.connect(addr).await.is_ok() {
+                            span_clone.in_scope(|| {
+                                debug!("service port detected open");
+                            });
+                            break;
+                        }
+                    }
+                    trace!("waiting for service port to open");
+                    // exponential backoff
+                    tokio::time::sleep(Duration::from_millis(ms)).await;
+                    ms *= 2;
+                }
+            })
+            .await;
+        }
```

Comment:
We can simplify the tracing a little bit, I quickly tested it locally and it's equivalent.
```suggestion
        if was_stopped {
            // wait until service has started and opens its port, give up after 10s
            span.in_scope(|| debug!("project waking up, checking service port"));
            let addr = SocketAddr::new(ip, DEPLOYER_SERVICE_HTTP_PORT);
            
            let _ = tokio::time::timeout(Duration::from_secs(10), async move {
                let mut ms = 5;
                loop {
                    if let Ok(socket) = TcpSocket::new_v4() {
                        if socket.connect(addr).await.is_ok() {
                                debug!("service port detected open");
                            break;
                        }
                    }
                    trace!("waiting for service port to open");
                    // exponential backoff
                    tokio::time::sleep(Duration::from_millis(ms)).await;
                    ms *= 2;
                }
            })
            .instrument(span.clone())
            .await;
        }
```